src/metabase/query_processor/middleware/resolve_fields.clj

```diff
@@ -16,7 +16,9 @@
   "Resolve all field referenced in the `query`, and store them in the QP Store."
   [query]
   (let [ids (into (set (mbql.u/match (:query query) [:field (id :guard integer?) _] id))
-                  (keep :id (get-in query [:query :source-metadata])))]
+                  (->> (mbql.u/match (:query query) {:source-metadata source-metadata} source-metadata)
```

Comment:
You are in a transducing context from the `into` so you could `(into (set ...) (comp cat (keep :id)) (mbql.u/match ...))` and skip the intermediate collections from the concat and keep.